cli/src/main/java/com/liferay/blade/cli/command/WatchCommand.java

```diff
@@ -75,102 +98,307 @@ private void _addError(String prefix, String msg) {
+							foundProjectPaths.put(_getGradlePath(path, watchPath), path);
+
+							return FileVisitResult.SKIP_SUBTREE;
+						}
 					}
+
+					return FileVisitResult.CONTINUE;
 				}
-			}
-		}
-		catch (IOException ioe) {
-		}
+
+			});
+
+		return foundProjectPaths;
 	}
 
-	private void _watch(Path watchPath, Map<String, Set<File>> projectOutputFiles) throws InterruptedException {
+	private void _registerDirectory(WatchService watcher, Map<WatchKey, Path> keys, Path dir) throws IOException {
+		WatchKey watchKey = dir.register(
+			watcher,
+			new WatchEvent.Kind[] {
+				StandardWatchEventKinds.ENTRY_CREATE, StandardWatchEventKinds.ENTRY_DELETE,
+				StandardWatchEventKinds.ENTRY_MODIFY
+			},
+			SensitivityWatchEventModifier.HIGH);
+
+		keys.put(watchKey, dir);
+	}
+
+	private void _walkAndRegisterDirectories(
+			final WatchService watchService, final Map<WatchKey, Path> watchKeys, final Path basePath,
+			final List<PathMatcher> ignorePathMatchers)
+		throws IOException {
+
+		Files.walkFileTree(
+			basePath,
+			new SimpleFileVisitor<Path>() {
+
+				@Override
+				public FileVisitResult preVisitDirectory(Path path, BasicFileAttributes basicFileAttributes)
+					throws IOException {
+
+					for (PathMatcher pathMatcher : ignorePathMatchers) {
+						if (pathMatcher.matches(path)) {
+							return FileVisitResult.SKIP_SUBTREE;
+						}
+					}
+
+					_registerDirectory(watchService, watchKeys, path);
+
+					return FileVisitResult.CONTINUE;
+				}
+
+			});
+	}
+
+	private void _watch(
+			Path watchPath, Map<String, Path> projectPaths, List<String> fastPaths, List<String> ignorePaths,
+			boolean deploy)
+		throws InterruptedException {
+
 		Thread watchThread = new Thread() {
 
 			@Override
 			public void run() {
 				BladeCLI bladeCLI = getBladeCLI();
 
 				try {
+					FileSystem fileSystem = FileSystems.getDefault();
+
+					final WatchService watchService = fileSystem.newWatchService();
+
+					final Map<WatchKey, Path> watchKeys = new HashMap<>();
+
+					final List<PathMatcher> ignorePathMatchers = _getPathMatchers(
+						watchPath, ignorePaths.toArray(new String[0]));
+
+					final List<PathMatcher> fastPathMatchers = _getPathMatchers(
+						watchPath, fastPaths.toArray(new String[0]));
+
+					_walkAndRegisterDirectories(watchService, watchKeys, watchPath, ignorePathMatchers);
+
 					final GradleExec gradleExec = new GradleExec(bladeCLI);
 
-					Set<String> projectPaths = projectOutputFiles.keySet();
+					if (deploy) {
+						bladeCLI.out("Deploying...  To skip initial deployment, use `blade watch -s`");
 
-					Stream<String> stream = projectPaths.stream();
+						gradleExec.executeTask("deploy", false);
+					}
 
-					String assembleTasks = stream.collect(Collectors.joining(":assemble "));
+					while (true) {
+						WatchKey watchKey;
 
-					String assembleTaskPath = assembleTasks + ":assemble";
+						try {
+							watchKey = watchService.take();
+						}
+						catch (InterruptedException ie) {
+							return;
```

Comment:
Perhaps `break` instead of `return`?

Reply:
Yeah i think we should break and keep the watch thread going.

Reply:
I'll do `continue` instead.